services/brig/test/integration/API/Settings.hs

```diff
@@ -0,0 +1,134 @@
+                ]
+            ]
+        ]
+
+data UserRelationship = SameTeam | DifferentTeam | NoTeam
+
+-- Should we show the email for this user type?
+type EmailVisibilityAssertion = UserRelationship -> Bool
+
+expectEmailVisible :: Opt.EmailVisibility -> UserRelationship -> Bool
+expectEmailVisible Opt.EmailVisibleIfOnTeam SameTeam = True
+expectEmailVisible Opt.EmailVisibleIfOnTeam DifferentTeam = True
+expectEmailVisible Opt.EmailVisibleIfOnTeam NoTeam = False
+
+expectEmailVisible Opt.EmailVisibleToSelf SameTeam = False
+expectEmailVisible Opt.EmailVisibleToSelf DifferentTeam = False
+expectEmailVisible Opt.EmailVisibleToSelf NoTeam = False
+
+jsonField :: FromJSON a => Text -> Value -> Maybe a
+jsonField f u = u ^? key f >>= maybeFromJSON
+
+testUsersEmailShowsEmailsIfExpected :: Brig -> Galley -> EmailVisibilityAssertion -> Http ()
+testUsersEmailShowsEmailsIfExpected brig galley shouldShowEmail = do
+    (creatorId, tid) <- createUserWithTeam brig galley
+    (otherTeamCreatorId, otherTid) <- createUserWithTeam brig galley
+    userA <- createTeamMember brig galley creatorId tid Team.fullPermissions
+    userB <- createTeamMember brig galley otherTeamCreatorId otherTid Team.fullPermissions
+    nonTeamUser <- createUser "joe" brig
+    let uids = C8.intercalate "," $ toByteString' <$> [userId userA, userId userB, userId nonTeamUser]
+        expected :: Set (Maybe UserId, Maybe Email)
+        expected = Set.fromList
+                   [ ( Just $ userId userA
+                     , if shouldShowEmail SameTeam then userEmail userA
+                                                   else Nothing)
+                   , ( Just $ userId userB
+                     , if shouldShowEmail DifferentTeam then userEmail userB
+                                                        else Nothing)
+                   , ( Just $ userId nonTeamUser
+                     , if shouldShowEmail NoTeam then userEmail nonTeamUser
+                                                 else Nothing)
+                   ]
+    get (brig . zUser (userId userB) . path "users" . queryItem "ids" uids) !!! do
+        const 200 === statusCode
+        const (Just expected) === result
+  where
+    result r =  Set.fromList
+             .  map (jsonField "id" &&& jsonField "email")
+            <$> decodeBody r
+
+testGetUserEmailShowsEmailsIfExpected :: Brig -> Galley -> EmailVisibilityAssertion -> Http ()
+testGetUserEmailShowsEmailsIfExpected brig galley shouldShowEmail = do
+    (creatorId, tid) <- createUserWithTeam brig galley
+    (otherTeamCreatorId, otherTid) <- createUserWithTeam brig galley
+    userA <- createTeamMember brig galley creatorId tid Team.fullPermissions
+    userB <- createTeamMember brig galley otherTeamCreatorId otherTid Team.fullPermissions
+    nonTeamUser <- createUser "joe" brig
+    let expectations :: [(UserId, Maybe Email)]
+        expectations =
```

Comment:
I personally prefer test code to be less redundant (but I think I am in a minority among people who write tests).

Also I haven't thought it through in this case.  It looks like it'd be easy to share more code, but I may well be wrong.

Reply:
It would be much easier to share code if we had a better 'fixtures' system in our tests; I've already collapsed about 4 tests down into 2 by allowing to pass the behaviour in, if I squish it down much further I fear it will hurt the already sketchy readability. If you have a concrete code suggestion I'll gladly accept it 😄